tests/unit/lms/views/api/canvas/sync_test.py

```diff
@@ -1,140 +1,148 @@
+):
+    groupids = sync(pyramid_request)
 
-    returned_groupids = sync(pyramid_request)
-
-    instructors_groups = [
-        factories.HGroup(
-            name="Section 1",
-            authority_provided_id="d0f36006728f2277f228b74c8a7f620305bfb3e7",
-            type="section_group",
-        ),
-        factories.HGroup(
-            name="Section 2",
-            authority_provided_id="d99674b9700f4a40a2b301d2949b61339c58236c",
-            type="section_group",
-        ),
-        factories.HGroup(
-            name="Section 3",
-            authority_provided_id="6a85e3651705dee4da0805b8985472343cbea94e",
-            type="section_group",
-        ),
-    ]
-    canvas_api_client.course_sections.assert_called_once_with(
-        "test_custom_canvas_course_id"
-    )
-    lti_h_service.sync.assert_called_once_with(instructors_groups, {"foo": "bar"})
-    canvas_api_client.users_sections.assert_called_once_with(
-        "test_canvas_user_id", "test_custom_canvas_course_id"
+    course_id = request_json["course"]["custom_canvas_course_id"]
+    user_id = request_json["learner"]["canvas_user_id"]
+    canvas_api_client.course_sections.assert_called_once_with(course_id)
+    canvas_api_client.users_sections.assert_called_once_with(user_id, course_id)
+
+    assert_sync_and_return(
+        groupids, synced_sections=sections.course, returned_sections=sections.user,
     )
-    assert returned_groupids == [
-        "group:d99674b9700f4a40a2b301d2949b61339c58236c@TEST_AUTHORITY"
-    ]
 
 
-def assert_that_it_called_sync_and_returned_the_groupids(
-    lti_h_service, returned_groupids, expected_groups
-):
-    lti_h_service.sync.assert_called_once_with(expected_groups, {"foo": "bar"})
-    assert returned_groupids == [
-        f"group:{group.authority_provided_id}@TEST_AUTHORITY"
-        for group in expected_groups
-    ]
+pytestmark = pytest.mark.usefixtures("canvas_api_client", "lti_h_service")
 
 
-pytestmark = pytest.mark.usefixtures("canvas_api_client", "lti_h_service")
+@pytest.fixture
+def assert_sync_and_return(lti_h_service, request_json):
+    tool_guid = request_json["lms"]["tool_consumer_instance_guid"]
+    context_id = request_json["course"]["context_id"]
+
+    def _groups_for(sections):
+        return [
+            HGroup.from_lti_parts(
+                name=section.get("name", f"Section {section['id']}"),
+                tool_consumer_instance_guid=tool_guid,
+                context_id=context_id,
+                section_id=section["id"],
+                type_="section_group",
+            )
+            for section in sections
+        ]
+
+    def assert_return_values(groupids, synced_sections, returned_sections):
+        lti_h_service.sync.assert_called_once_with(
+            _groups_for(synced_sections), request_json["group_info"]
+        )
+
+        assert groupids == [
+            group.groupid(TEST_SETTINGS["h_authority"])
+            for group in _groups_for(returned_sections)
+        ]
+
```

Comment:
A caller pattern fixture which has a closure around a closure has to be some kind of `pytest` bingo win.